zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

```diff
@@ -0,0 +1,612 @@
+        // Calculate transaction gas fee
+        let mul_gas_fee_by_gas = MulWordByU64Gadget::construct(
+            cb,
+            tx_gas_price.clone(),
+            tx_gas.clone(),
+            true,
+        );
+
+        // TODO: Take gas cost of access list (EIP 2930) into consideration.
+        // Use intrinsic gas
+        let intrinsic_gas_cost =
+            select::expr(tx_is_create.expr(), 53000.expr(), 21000.expr())
+                + tx_call_data_gas_cost.expr();
+
+        // Check gas_left is sufficient
+        let gas_left = tx_gas.expr() - intrinsic_gas_cost;
+        let sufficient_gas_left =
+            RangeCheckGadget::construct(cb, gas_left.clone());
+
+        // Prepare access list of caller and callee
+        cb.account_access_list_write(
+            tx_id.expr(),
+            tx_caller_address.expr(),
+            1.expr(),
+            0.expr(),
+        );
+        cb.account_access_list_write(
+            tx_id.expr(),
+            tx_callee_address.expr(),
+            1.expr(),
+            0.expr(),
+        );
+
+        // Transfer value from caller to callee
+        let transfer_with_gas_fee = TransferWithGasFeeGadget::construct(
+            cb,
+            tx_caller_address.expr(),
+            tx_callee_address.expr(),
+            tx_value.clone(),
+            mul_gas_fee_by_gas.product().clone(),
+            is_persistent.expr(),
+            rw_counter_end_of_reversion.expr(),
+        );
+
+        // Assume it's not a creation transaction nor calling a precompiled
+        cb.add_constraint(
+            "Handling of creation transaction has yet to be determined",
+            tx_is_create.expr(),
+        );
+
+        // Read code_hash of callee
+        let code_hash = cb.query_cell();
+        cb.account_read(
+            tx_callee_address.expr(),
+            AccountFieldTag::CodeHash,
+            code_hash.expr(),
+        );
+
+        // Setup next call's context.
+        for (field_tag, value) in [
+            (CallContextFieldTag::Depth, 1.expr()),
+            (CallContextFieldTag::CallerAddress, tx_caller_address.expr()),
+            (CallContextFieldTag::CalleeAddress, tx_callee_address.expr()),
+            (CallContextFieldTag::CallDataOffset, 0.expr()),
+            (
+                CallContextFieldTag::CallDataLength,
+                tx_call_data_length.expr(),
+            ),
+            (CallContextFieldTag::Value, tx_value.expr()),
+            (CallContextFieldTag::IsStatic, 0.expr()),
+        ] {
+            cb.call_context_lookup(Some(call_id.expr()), field_tag, value);
+        }
+
+        cb.require_step_state_transition(StepStateTransition {
+            rw_counter: Delta(16.expr()),
+            call_id: To(call_id.expr()),
+            is_root: To(1.expr()),
+            is_create: To(0.expr()),
+            opcode_source: To(code_hash.expr()),
+            program_counter: To(0.expr()),
+            stack_pointer: To(1024.expr()),
```

Comment:
Probably using a `const` for the stack len would be better as mentioned avobe.

Reply:
Fixed in [`bb2f101`](https://github.com/appliedzkp/zkevm-circuits/pull/230/commits/bb2f1012ec8dd8b6d5f91f87bba2757180fdef60).